test/corecel/random/Histogram.hh

```diff
@@ -39,39 +42,79 @@ class Histogram
     // Update the histogram with a value
     inline void operator()(double value);
 
+    // Update the histogram with a vector of values
+    inline void operator()(VecDbl const& values);
+
     // Get the histogram
     VecCount const& counts() const { return counts_; }
 
     // Get the result as a probability density
     VecDbl calc_density() const;
 
+    // Get the overflow and overflow counts
+    size_type underflow() const { return out_of_range_[Bound::lo]; }
+    size_type overflow() const { return out_of_range_[Bound::hi]; }
+
+    // Get the minimum and maximum values encountered
+    double min() const { return extrema_[Bound::lo]; }
+    double max() const { return extrema_[Bound::hi]; }
+
   private:
+    using Limits = std::numeric_limits<double>;
+
     double offset_;
     double inv_width_;
     VecCount counts_;
-    size_type total_counts_{0};
+
+    // Keep track of underflow and overflow
+    EnumArray<Bound, size_type> out_of_range_{};
+    EnumArray<Bound, double> extrema_{Limits::infinity(), -Limits::infinity()};
 };
 
 //---------------------------------------------------------------------------//
 /*!
  * Update the histogram with a value.
  *
- * Values outside of \c range are allowable and will show as a deficit in the
- * resulting tally.
+ * Values outside of \c range are ignored. All bins are half-open except for
+ * the rightmost bin, which will include values equal to the upper edge.
  */
 void Histogram::operator()(double value)
 {
-    ++total_counts_;
     double frac = (value - offset_) * inv_width_;
-    if (frac < 0.0 || frac >= 1.0)
+    if (frac < 0.0)
     {
+        ++out_of_range_[Bound::lo];
+        extrema_[Bound::lo] = std::min(extrema_[Bound::lo], value);
+        return;
+    }
+    else if (frac > 1.0)
```

Comment:
Minor thing to watch out for: it's a bit safer to order this as `if (frac < 0) else frac < 1 else frac == 1 else ...`. This way, NaN (which is false for all comparisons) falls into the upper extremum and it would stay there. As written it's dangerous because `static_cast<size_type>(NaN)` is UB.

Reply:
Good point!